doc/rst/developer/chips/23.rst

```diff
@@ -0,0 +1,614 @@
+to any function.  However any computation involving local variables
+and module level procedures is permitted. The developer may skip a
+field and rely on the compiler to insert the default value defined by
+the field declaration.
+
+We introduce a new method, currently named postInit(), to support an
+explicit phase 2 when required.  The implementation of a new
+expression will arrange to call postInit() when the selected
+init(..args..) returns i.e. we consider
+
+.. code-block:: chapel
+
+   var x = new MyRecord(...)
+
+to be equivalent to
+
+.. code-block:: chapel
+
+   var x : MyRecord = noinit;
+
+   x.init(...args...);
+   x.postInit();
+
+
+On entry to postInit, the instance is known to be a fully initialized
+instance of MyRecord. The developer may override this method and
+implement any code to customize the instance.
+
+There will be cases in which a record does not implement any overrides
+for init() and the compiler will provide a default all-fields
+initializer.  The provision of postInit() provides the same post-init
+hook that is currently supported by the initialize() method for types
+that rely on legacy constructors.
+
+It is observed that the separation between init() and postInit()
+might be viewed as either "clean" or "overly constraining"
+depending on taste.  It is unclear how common it will
+be for users to implement both init() and postInit() and
+how different users will feel when obliged to do so.
+
+In the longer term we might choose to consider the common
+cases that lead to an override for postInit() and then
+make backward-compatible changes that provide some
+additional flexibility within init().
+
+
+
+
+
+
+Base Classes
+++++++++++++
+
+Here we define a Base Class as a class that is derived from object.
+Currently the type 'object' does not have any user-visible fields and
+very little behavior.  We take the perspective that base classes are
+comparable to records as far as initialization is concerned.  This
+implies that init() has the functionality of phase 1 in the current
+implementation, and that postInit() has the functionality of phase 2
+in the current implementation.  Of course a method dispatch within
+postInit() will dynamically dispatch to the runtime type of the
+instance.
+
+
+
+
+Derived Classes
++++++++++++++++
+
+This is the case where the tradeoffs become more challenging.  The
+intent is avoid passing an object with uninitialized fields, and to
+avoid unpredictable/confusing interactions between components of the
+distributed implementation for the overall initializer.
+
+One choice is to allow derived classes to override one of both
+of init() and postInit() without any further extensions.  Additionally
+Derived.postInit() would not explictly invoke Parent.postInit(); the
+implementation would arrange to work "as if" from base class towards
+most derived class to ensure that a derived class can "own"
+the final value of every visible field.
```

Comment:
I'm having a hard time parsing the last part of this sentence -- to work as if *what* from base class towards most derived?  (Though I gather it means the ``postInit()``s are called in that order.)